src/connect/selectorFactory.ts

```diff
@@ -163,24 +150,34 @@ export function pureFinalPropsSelectorFactory<
   }
 }
 
+interface WrappedMapStateToProps<TStateProps, TOwnProps, State> {
+  (state: State, ownProps: TOwnProps): TStateProps
+  readonly dependsOnOwnProps: boolean
+}
+
+interface WrappedMapDispatchToProps<TDispatchProps, TOwnProps> {
+  (dispatch: Dispatch<Action<unknown>>, ownProps: TOwnProps): TDispatchProps
+  readonly dependsOnOwnProps: boolean
+}
+
 export interface SelectorFactoryOptions<
   TStateProps,
   TOwnProps,
   TDispatchProps,
   TMergedProps,
-  State = unknown
-> extends PureSelectorFactoryComparisonOptions<TOwnProps, State> {
+  State
+> extends PureSelectorFactoryComparisonOptions<TStateProps, TOwnProps, State> {
   initMapStateToProps: (
     dispatch: Dispatch,
-    options: PureSelectorFactoryComparisonOptions<TOwnProps, State>
-  ) => MapStateToPropsParam<TStateProps, TOwnProps, State>
+    options: PureSelectorFactoryComparisonOptions<TStateProps, TOwnProps, State>
+  ) => WrappedMapStateToProps<TStateProps, TOwnProps, State>
   initMapDispatchToProps: (
     dispatch: Dispatch,
-    options: PureSelectorFactoryComparisonOptions<TOwnProps, State>
-  ) => MapDispatchToPropsParam<TDispatchProps, TOwnProps>
+    options: PureSelectorFactoryComparisonOptions<TStateProps, TOwnProps, State>
+  ) => WrappedMapDispatchToProps<TDispatchProps, TOwnProps>
```

Comment:
This is the main change. `initMapStateToProps` and `initMapDispatchToProps` have already been sanitized and conformed to a single function type, so using `MapStateToPropsParam` and `MapDispatchToPropsParam` were unnecessarily broad and the source of most of the type errors.